File: experiments/006-SmalltalkVsJava/cells/1-Evaluate+TestRunning-University_1-Empty_free/2024-1c-parcial1/20260915-112638-41597/exercise/java/src/main/java/simshotels/Room.java
package simshotels;

import java.util.Collections;
import java.util.HashMap;
import java.util.function.Supplier;

public class Room {

    private static final String RESERVED = "reserved";

    private String guest;

    // guests

    public String guestType() {
        return guest;
    }

    public void receive(String aGuestType) {
        // Room Occupied
        if (guest != null || RESERVED.equals(guest)) {
            throw new RuntimeException(Room.roomIsNotEmptyErrorDescription());
        } else {
            guest = aGuestType;
        }
    }

    public void receiveWithReservation(String aGuestType) {
        // Room Reserved
        if (RESERVED.equals(guest)) {
            guest = aGuestType;
        } else {
            throw new RuntimeException(Room.roomIsNotReservedErrorDescription());
        }
    }

    public void reserve() {
        // Room Occupied
        if (guest != null || RESERVED.equals(guest)) {
            throw new RuntimeException(Room.roomIsNotEmptyErrorDescription());
        } else {
            guest = RESERVED;
        }
    }

    // testing

    public boolean isAvailable() {
        return guest == null;
    }

    public boolean isOccupied() {
        return guest != null || RESERVED.equals(guest);
    }

    public boolean isReserved() {
        return RESERVED.equals(guest);
    }

    // accounting

    public Integer profitUsingIfAbsentGuestType(HashMap<String, Integer> aPriceList, Supplier<Integer> unknownGuestTypeBlock) {
        // Room Available
        if (guest == null) {
            return 0;
        }

        // Room Reserved
        if (RESERVED.equals(guest)) {
            return Collections.min(aPriceList.values()) / 2;
        }

        // Room Occupied
        if (guest != null || RESERVED.equals(guest)) {
            if (aPriceList.containsKey(guest)) {
                return aPriceList.get(guest);
            } else {
                return unknownGuestTypeBlock.get();
            }
        }

        return null;
    }

    public Integer lossUsingIfAbsentGuestType(HashMap<String, Integer> aPriceList, Supplier<Integer> unknownGuestTypeBlock) {
        // Not implemented yet: in the Smalltalk original this message does not exist in Room
        // (sending it raises MessageNotUnderstood). Java needs the method to compile the tests.
        throw new UnsupportedOperationException("Room does not understand lossUsingIfAbsentGuestType");
    }

    // error messages (class side)

    public static String roomIsNotEmptyErrorDescription() {
        return "Room is not empty.";
    }

    public static String roomIsNotReservedErrorDescription() {
        return "Room is not reserved.";
    }
}

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260921-215532-49821/exercise/python/src/simshotels/room.py
class Room:

    RESERVED = "reserved"

    def __init__(self):
        self._guest = None

    # guests

    def guest_type(self):
        return self._guest

    def receive(self, a_guest_type):
        # Room Occupied
        if self._guest is not None or Room.RESERVED == self._guest:
            raise RuntimeError(Room.room_is_not_empty_error_description())
        else:
            self._guest = a_guest_type

    def receive_with_reservation(self, a_guest_type):
        # Room Reserved
        if Room.RESERVED == self._guest:
            self._guest = a_guest_type
        else:
            raise RuntimeError(Room.room_is_not_reserved_error_description())

    def reserve(self):
        # Room Occupied
        if self._guest is not None or Room.RESERVED == self._guest:
            raise RuntimeError(Room.room_is_not_empty_error_description())
        else:
            self._guest = Room.RESERVED

    # testing

    def is_available(self):
        return self._guest is None

    def is_occupied(self):
        return self._guest is not None or Room.RESERVED == self._guest

    def is_reserved(self):
        return Room.RESERVED == self._guest

    # accounting

    def profit_using_if_absent_guest_type(self, a_price_list, unknown_guest_type_block):
        # Room Available
        if self._guest is None:
            return 0

        # Room Reserved
        if Room.RESERVED == self._guest:
            return min(a_price_list.values()) // 2

        # Room Occupied
        if self._guest is not None or Room.RESERVED == self._guest:
            if self._guest in a_price_list:
                return a_price_list[self._guest]
            else:
                return unknown_guest_type_block()

    # error messages (class side)

    @staticmethod
    def room_is_not_empty_error_description():
        return "Room is not empty."

    @staticmethod
    def room_is_not_reserved_error_description():
        return "Room is not reserved."
